coalib/misc/Caching.py

```diff
@@ -0,0 +1,122 @@
+import calendar
+import time
+import hashlib
+
+from coalib.misc.CachingUtilities import (
+    pickle_load, pickle_dump, time_consistent, update_time_db,
+    delete_cache_files)
+
+
+class FileCache:
+
+    def __init__(self, log_printer, project_dir, flush_cache=False):
+        """
+        Initialize cache instance for project. Example usage:
+
+        >>> from pyprint.NullPrinter import NullPrinter
+        >>> from coalib.output.printers.LogPrinter import LogPrinter
+        >>> import copy, time
+        >>> log_printer = LogPrinter(NullPrinter())
+
+        To initialize the cache create an instance for the project:
+        >>> cache = FileCache(log_printer, "coala_test", flush_cache=True)
+
+        Now we can track new files by running:
+        >>> cache.track_new_files(["a.c", "b.c"])
+
+        Since all cache operations are lazy (for performance), we need to
+        explicitly write the cache to disk for persistence in future uses:
+        (Note: The cache will automatically figure out the write location)
+        >>> cache.write()
+
+        >>> time.sleep(1)
```

Comment:
Not all file systems support fractional file stat times. So we need to necessarily floor the stat time to be consistent. This makes the `sleep(1)` necessary to simulate future.